### PR TITLE
Use existing array for precompile result

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Use existing array for precompile result, when Memory array backed and size same

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/Precompiles/IdentityPrecompile.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public long DataGasCost(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseS
 
         public (ReadOnlyMemory<byte>, bool) Run(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec)
         {
-            return (inputData, true);
+            return (inputData.ToArray(), true);
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Evm/Precompiles/IdentityPrecompile.cs:32]: Not enturely sure why, but IdentityPrecompile isn't really used so shouldn't really matter

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 2]
```diff
@@ -690,7 +691,13 @@ private CallResult ExecutePrecompile(EvmState state, IReleaseSpec spec)
...
         {
             (ReadOnlyMemory<byte> output, bool success) = precompile.Run(callData, spec);
-            CallResult callResult = new(output.ToArray(), success, !success);
+            byte[] result = (MemoryMarshal.TryGetArray(output, out ArraySegment<byte> outputArray) &&
+                outputArray.Offset == 0 && outputArray.Count == output.Length) ?
+                outputArray.Array :
+                // Wasn't array backed or Memory wasn't the whole array
+                output.ToArray();
+
+            CallResult callResult = new(result, success, !success);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:700]: We can just change `CallResult` to accept `ReadOnlyMemory<byte>` as it is already like this in property:

```
public ReadOnlyMemory<byte> Output { get; }
```

**Review by LukaszRozmej — COMMENTED** [review]
Also would it be better to be `bool TryRun(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec, out ReadOnlyMemory<byte> result)` ?

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs]: Changed

**Comment by benaadams** [comment]
> Also would it be better to be `bool TryRun(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec, out ReadOnlyMemory<byte> result)` ?

More of a break if another chain has their own precompiles? (Don't know if they do)

**Comment by LukaszRozmej** [comment]
> > Also would it be better to be `bool TryRun(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec, out ReadOnlyMemory<byte> result)` ?
> 
> More of a break if another chain has their own precompiles? (Don't know if they do)

huh why? I am askung about just changing interface, `bool` can go via register, or the tuple goes via registers?

**Review by LukaszRozmej — APPROVED** [review]

**Comment by benaadams** [comment]
> > > Also would it be better to be `bool TryRun(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec, out ReadOnlyMemory<byte> result)` ?
> > 
> > 
> > More of a break if another chain has their own precompiles? (Don't know if they do)
> 
> huh why? I am askung about just changing interface, `bool` can go via register, or the whole tuple goes via registers?

I mean if any existing extenal precompiles/plugin as they would need to change their signiture to upgrade?

**Comment by benaadams** [comment]
Would be a better signiture though :)

**Comment by LukaszRozmej** [comment]
> > > > Also would it be better to be `bool TryRun(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec, out ReadOnlyMemory<byte> result)` ?
> > > 
> > > 
> > > More of a break if another chain has their own precompiles? (Don't know if they do)
> > 
> > 
> > huh why? I am askung about just changing interface, `bool` can go via register, or the whole tuple goes via registers?
> 
> I mean if any existing extenal precompiles/plugin as they would need to change their signiture to upgrade?

I don't think there are any, and if yes this is trivial change

**Pull request merged**